opentelemetry-otlp/src/exporter/http/mod.rs

```diff
@@ -405,11 +378,64 @@ fn add_header_from_string(input: &str, headers: &mut HashMap<HeaderName, HeaderV
     }));
 }
 
+/// Expose interface for modifying builder config.
+pub trait HasHttpConfig {
```

Comment:
Why are these new traits `HasHttpConfig` and `WithHttpConfig` needed? Could we have these methods as part of `HttpExporterBuilder like earlier`?

Reply:
Those are so the config options are available on the Span / Metric / Log Exporter builders when with_http is used

Reply:
If they are on the HttpExporterBuilder we would have to take that as a parameter for the with_http

Reply:
I don't follow. How does `with_http` influence `with_http_client`?
- `with_http` would bring in the default client (reqwest)
- `with_http_client` allows the user to bring their custom client

Both of these methods can reside in `impl HttpExporterBuilder`. What value are these new public traits adding?

Reply:
for example, using the new traits the user gains access to the http specific config options without needing to configure a separate `HttpExporterBuilder` / `TonicExporterBuilder`. So in the following case, after `.with_http()` the two http specific methods become available, but the tonic specific ones aren't.
```rust
let exporter = SpanExporter::builder()
    .with_http()
    .with_http_client(...) // non default client like hyper
    .with_headers(...) // non default headers to send
    .build()?;

let exporter = SpanExporter::builder()
    .with_tonic()
    .with_metadata(...) // non default metadata to send
    .build()?;
```

Without the traits we would have to do something like:
```rust
let exporter = SpanExporter::builder()
    .with_http(HttpExporter::builder().with_headers(...).build())
    .build()?;
```

Reply:
>How does `with_http` influence `with_http_client`?

`with_http()` alone will use the default values for http_client / headers, but after you say `with_http()` you can configure that http exporter with non default values like `.with_http_client(hyper)` to set a non default client before you call `build()`

Reply:
To unblock this PR (its pretty big and will cause merge conflicts if delayed more), lets track this separately as a follow up ?

As per the SIG/Community call yesterday, we are delaying 0.27 release by a week, so we can do follow ups as necessary after merge. Also, I think changes to this area might be less intrusive to end users (not 100% sure).

@lalitb @utpilla Please share your thoughts on merging this as-is and follow up on these extra traits afterwards?

Reply:
Yes, this is good discussion. Traits are not affecting the API surface, so can be revisited aftewards.

Reply:
This PR brings a lot of new public API additions:
- `pub struct HttpConfig`
- `pub trait HasHttpConfig`
- `pub trait WithHttpConfig`
- `pub trait HasTonicConfig`
- `pub trait WithTonicConfig`
- `pub struct NoExporterBuilderSet`
- `pub struct TonicExporterBuilderSet(TonicExporterBuilder)`
- pub struct HttpExporterBuilderSet(HttpExporterBuilder)`

I would like us to achieve this with a reduce public API surface. I haven't looked at this deeply to assess whether it's feasible or not.

I also agree that even though these are public, we don't expect users to be using them. However, a reduced public API surface generally eases the maintenance burden in the long run. I would prefer that we attempt to simplify this if we are merging it now. My question would be if that simplification process demands any breaking change in the usage (especially setup of the exporter), would that be acceptable?

Reply:
> My question would be if that simplification process demands any breaking change in the usage (especially setup of the exporter), would that be acceptable?

In my opinion -  Yes, given SDK/OTLP Exporter crates are still going to be alpha/beta for at-least few more weeks.

@pitoniak32 We can merge this as-is now. Would you have some bandwidth to see if this can be further simplified as a follow up?

Reply:
Yeah I can look more into this, I won't be able to get started until tomorrow, my main desktop motherboard needs to be RMAd because of memory errors :( so I need to setup my dev env on another computer but I should still be able to look into it

Reply:
> @pitoniak32 We can merge this as-is now. Would you have some bandwidth to see if this can be further simplified as a follow up?

Should we have a discussion around what would be ideally included in the new solution? I joined the otel rust slack if you would like to chat there, or would it be better to open a discussion / issue? I can also just jump into trying alternatives and open a draft PR. But it could be good to align on what would be a nice interface to configure these exporters

Reply:
Thanks @pitoniak32! I'll reach out to you on Slack with my initial thoughts.

Reply:
@utpilla okay to merge now?
(Trying to avoid delaying big PRs to avoid conflicts)

Reply:
Yes.